src/exec/execute_command.c

```diff
@@ -58,6 +58,10 @@ static void	restore_std_io(int *std_fds)
 	close(std_fds[2]);
 }
 
+/*
+** argv is NULL only if no arguments are passed to the command.
+** In this case, this function returns EXIT_SUCCESS.
+*/
 static int	execute_builtin_command(t_ast *node, t_ctx *ctx)
```

Comment:
radical suggestion to use return early pattern.

```c
static int	execute_builtin_command(t_ast *node, t_ctx *ctx)
{
	const char	*cmd_name;
	char		**argv;
	int			status;

	status = 0;
	cmd_name = get_cmd_arg(node->cmd_args);
	argv = convert_cmd_args_to_array(node->cmd_args);
	if (ft_strcmp(cmd_name, "cd") == 0)
		status = builtins_cd(argv, ctx);
	else if (ft_strcmp(cmd_name, "pwd") == 0)
		status = builtins_pwd(argv, ctx);
	else if (ft_strcmp(cmd_name, "echo") == 0)
		status = builtins_echo(argv, ctx);
	else if (ft_strcmp(cmd_name, "env") == 0)
		status = builtins_env(argv, ctx);
	else if (ft_strcmp(cmd_name, "export") == 0)
		status = builtins_export(argv, ctx);
	else if (ft_strcmp(cmd_name, "unset") == 0)
		status = builtins_unset(argv, ctx);
	else if (ft_strcmp(cmd_name, "exit") == 0)
		status = builtins_exit(argv, ctx);
	ft_free_strs(argv);
	return (status);
}

int	execute_command(t_ast *node, t_ctx *ctx, t_pipeline_conf *conf)
{
	int		std_fds[3];
	t_list	*expanded_cmd_args;

	if (node->cmd_args == NULL)
	{
		ctx->exit_status = EXIT_SUCCESS;
		return (EXIT_SUCCESS);
	}
	expanded_cmd_args = expand(node->cmd_args, ctx);
	ft_lstclear(&node->cmd_args, free);
	node->cmd_args = expanded_cmd_args;
	if (is_builtin(get_cmd_arg(node->cmd_args)))
	{
		save_std_io(std_fds);
		if (handle_io(conf, node->redirects, ctx))
			ctx->exit_status = execute_builtin_command(node, ctx);
		else
			ctx->exit_status = EXIT_FAILURE;
		restore_std_io(std_fds);
		return (EXIT_SUCCESS);
	}
	else
		return (spawn_command(node, ctx, conf));
}
```

Reply:
So `handle_io` must always be applied.